evalml/automl/engines/engine_base.py

```diff
@@ -0,0 +1,184 @@
+from evalml.pipelines import BinaryClassificationPipeline
+from evalml.problem_types import ProblemTypes
+from evalml.utils.gen_utils import (
+    _convert_to_woodwork_structure,
+    _convert_woodwork_types_wrapper
+)
+from evalml.utils.logger import get_logger, update_pipeline
+
+logger = get_logger(__file__)
+
+
+class EngineBase(ABC):
+    def __init__(self):
+        self.name = "Base Engine"
+        self.X = None
+        self.y = None
+        self.automl = None
+
+    def load_data(self, X, y):
+        self.X = X
+        self.y = y
+
+    def load_search(self, search_obj):
+        self.automl = search_obj
+
+    @abstractmethod
+    def evaluate_batch(self, pipeline_batch=None):
+        if self.X is None or self.y is None:
+            raise ValueError("Dataset has not been loaded into the engine. Call `load_data` with training data.")
+
+        if self.automl is None:
+            raise ValueError("Search info has not been loaded into the engine. Call `load_search` with search context.")
+
+    @abstractmethod
+    def evaluate_pipeline(self, pipeline=None):
+        if self.X is None or self.y is None:
+            raise ValueError("Dataset has not been loaded into the engine. Call `load_data` with training data.")
+
+        if self.automl is None:
+            raise ValueError("Search info has not been loaded into the engine. Call `load_search` with search context.")
+
+    def log_pipeline(self, pipeline, current_iteration=None):
+        desc = f"{pipeline.name}"
+        if len(desc) > self.automl._MAX_NAME_LEN:
+            desc = desc[:self.automl._MAX_NAME_LEN - 3] + "..."
+        desc = desc.ljust(self.automl._MAX_NAME_LEN)
+
+        update_pipeline(logger,
+                        desc,
+                        (current_iteration if current_iteration else len(self.automl._results['pipeline_results'])) + 1,
+                        self.automl.max_iterations,
+                        self.automl._start,
+                        1 if self.automl._automl_algorithm.batch_number == 0 else self.automl._automl_algorithm.batch_number,
+                        self.automl.show_batch_output)
+
+    def _add_result_callback(self, result_callback, pipeline, evaluation_result):
+        parameters = pipeline.parameters
+        logger.debug('Adding results for pipeline {}\nparameters {}\nevaluation_results {}'.format(pipeline.name, parameters, evaluation_result))
+        result_callback(trained_pipeline=pipeline,
+                        parameters=parameters,
+                        training_time=evaluation_result['training_time'],
+                        cv_data=evaluation_result['cv_data'],
+                        cv_scores=evaluation_result['cv_scores'])
+        logger.debug('Adding results complete')
+
+        if self.automl.search_iteration_plot:
+            self.automl.search_iteration_plot.update()
+
+    def _handle_keyboard_interrupt(self, current_batch_pipelines, pipeline=None):
```

Comment:
The keyboard interrupt works, but I think there's still some room to tweak this.

Reply:
The behavior is more or less the same. However, in the context of engines, the mechanism of emptying out the batch on termination feels unnecessary. I might put some more thought in what a cleaner implementation would look like but this one currently works as-is.